course/page/code.py

```diff
@@ -292,6 +294,97 @@ def request_python_run_with_retries(run_req, run_timeout, image=None, retry_coun
         return result
 
 
+# {{{ HTML sanitization helpers
+
+def is_allowed_data_uri(allowed_mimetypes, uri):
+    import re
+    m = re.match(r"^data:([-a-z0-9]+/[-a-z0-9]+);base64,", uri)
+    if not m:
+        return False
+
+    mimetype = m.group(1)
+    return mimetype in allowed_mimetypes
+
+
+def filter_from_code_html_attributes(tag, name, value):
+    from html5lib.filters.sanitizer import attr_val_is_uri
+    if tag == "audio":
+        if name in ["controls"]:
+            return True
+        else:
+            return False
+
+    elif tag == "source":
+        if name in ["type"]:
+            return True
+        elif name == "src":
+            if is_allowed_data_uri([
+                    "audio/wav",
+                    ], value):
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    elif tag == "img":
+        if name in ["alt", "title"]:
+            return True
+        elif name == "src":
+            if is_allowed_data_uri([
+                    "image/png",
+                    "image/jpeg",
+                    ], value):
+                return True
+        else:
+            return False
+
+    else:
```

Comment:
IMO, the logic below is redudant. After proceeding the logic below, bleach will do the same logic again. IOW, We are still [here](https://github.com/mozilla/bleach/blob/8706b5373b633407e804b1b2975edf8760067ff7/bleach/sanitizer.py#L508) after finishing the whole function.

The problem is, we didn't filter those built-in allowed tags with protocol allowed, in the logic above. So `<style src="data:,Evilcss">` just get passed.